File: PyUQTk/utils/func.py
#!/usr/bin/env python
#=====================================================================================
#
#                      The UQ Toolkit (UQTk) version 3.1.2
#                          Copyright (2022) NTESS
#                        https://www.sandia.gov/UQToolkit/
#                        https://github.com/sandialabs/UQTk
#
#     Copyright 2022 National Technology & Engineering Solutions of Sandia, LLC (NTESS).
#     Under the terms of Contract DE-NA0003525 with NTESS, the U.S. Government
#     retains certain rights in this software.
#
#     This file is part of The UQ Toolkit (UQTk)
#
#     UQTk is open source software: you can redistribute it and/or modify
#     it under the terms of BSD 3-Clause License
#
#     UQTk is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     BSD 3 Clause License for more details.
#
#     You should have received a copy of the BSD 3 Clause License
#     along with UQTk. If not, see https://choosealicense.com/licenses/bsd-3-clause/.
#
#     Questions? Contact the UQTk Developers at <uqtk-developers@software.sandia.gov>
#     Sandia National Laboratories, Livermore, CA, USA
#=====================================================================================

"""
Generic tools for evaluation of standard functions
and their integrals
"""

try:
    import numpy as np
except ImportError:
    "Need numpy"

import sys,os
from math import *
import random as rnd
import itertools


from .mindex_order import getNPC

###################################################################################################


def func(xdata,model,func_params):
    """Generic function evaluator.
    Note:
        * Note that conventional Genz arguments are in [0,1], here the expected input is on [-1,1]
               The shifts are applied to the transformed coordinates in [0,1], to be consistent
               with traditional formulations of Genz functions.
               In future versions of UQTk, we may redefine the Genz functions here to take
               arguments in [0,1]. Need to check if that will create trouble with the other
               function types defined here. Maybe take Genz functions out and make them
               a separate functionality?
    Arguments:
        * xdata       : Nxd numpy array of input, should be in [-1,1]^d
        * model       : Model name, options are 'genz_osc', 'genz_exp', 'genz_cont', 'genz_gaus',
                        'genz_cpeak', 'genz_ppeak', 'ishigami', 'sobol', 'poly_exsens', 'PCmi', 'spp'
        * func_params : Auxiliary parameters
                      : For genz functions, an array of size d+1, the first entry being the shift,
                      : which is the same for all dimensions.
                      : and the rest of the entries are the weights.
                      : See UQTk Manual for Genz formulae.
    Returns:
        * ydata       : An array of outputs of size N.
    """

    # Get the input size
    sam=xdata.shape[0]
    dim=xdata.shape[1]

    # Check the function types and evaluate
    if model == 'genz_osc':
        xdata=0.5*(xdata+1.)
        ydata=np.empty((sam,))
        gcf=func_params[1:]
        xtmp=np.dot(xdata,gcf)
        for j in range(sam):
            ydata[j]=cos(2.*pi*func_params[0]+xtmp[j])

    elif model == 'genz_exp':
        xdata=0.5*(xdata+1.)
        ydata=np.empty((sam,))
        ww=func_params[0]
        gcf=func_params[1:]
        print('The shape of ww is: ',ww.shape)
        print('The shape of gcf is: ',gcf.shape)
        print('The shape of x is: ',xdata.shape)
        xtmp=np.dot(xdata-ww,gcf)
        for j in range(sam):
            ydata[j]=exp(xtmp[j])

    elif model == 'genz_cont':
        xdata=0.5*(xdata+1.)
        ydata=np.empty((sam,))
        ww=func_params[0]
        gcf=func_params[1:]

        xtmp=np.dot(abs(xdata-ww),gcf)
        for j in range(sam):
            ydata[j]=exp(-xtmp[j])

    elif model == 'genz_gaus':
        xdata=0.5*(xdata+1.)
        ydata=np.empty((sam,))
        ww=func_params[0]
        gcf=func_params[1:]

        xtmp=np.dot((xdata-ww)*(xdata-ww),gcf*gcf)
        for j in range(sam):
            ydata[j]=exp(-xtmp[j])

    elif model == 'genz_cpeak':
        xdata=0.5*(xdata+1.)
        ydata=np.empty((sam,))
        #ww=param[0]
        gcf=func_params[1:]

        xtmp=1.+(np.dot(xdata,gcf)) #use abs if defined on [-1,1]
        for j in range(sam):
            ydata[j]=exp(-(dim+1.)*log(xtmp[j]))

    elif model == 'genz_ppeak':
        xdata=0.5*(xdata+1.)
        ydata=np.empty((sam,))
        ww=func_params[0]
        gcf=func_params[1:]

        for j in range(sam):
            prod=1.
            for i in range(dim):
                prod = prod / (1./(gcf[i]**2.)+(xdata[j,i]-ww)**2.)
            ydata[j]=prod

    elif model == 'ishigami':
        assert(dim==3)
        a=func_params[0]
        b=func_params[1]
        ydata=np.empty((sam,))

        for j in range(sam):
            ydata[j]=np.sin(xdata[j,0])+a*np.sin(xdata[j,1])**2+b*np.sin(xdata[j,0])*xdata[j,2]**4

    elif model == 'sobol':
        assert(dim==func_params.shape[0])
        ydata=np.empty((sam,))
        for j in range(sam):
            val=1.
            for k in range(dim):
                val *= ( (abs(2*xdata[j,k])+func_params[k])/(1.+func_params[k]) )
            ydata[j]=val

    elif model == 'poly_exsens':
        assert(dim==func_params[0])
        ydata=np.empty((sam,))
        for j in range(sam):
            val=1.
            for k in range(dim):
                val *= ( (3./4.)*(xdata[j,k]+1.)**2+1. )/2.
            ydata[j]=val

    elif model == 'PCmi':
        mindex=func_params[0]
        pccf=func_params[1]
        pctype=func_params[2]

        npc=mindex.shape[0]

        #print "mindex=",mindex, pccf

        np.savetxt('xdata.dat',xdata)

        np.savetxt('pccf.dat',pccf)
        np.savetxt('mi.dat',mindex,fmt='%d')
        assert npc==pccf.shape[0]


        # Evaluate PC to get the training data
        if pctype=='MON':
            ydata=np.zeros((sam,))
            for ipt in range(sam):
                for i in range(npc):
                    cur=pccf[i]
                    for j in range(dim):
                        cur*=(xdata[ipt,j]**mindex[i,j])
                    ydata[ipt]+=cur
        else:
            cmd='pce_eval -x"PC_mi" -f"pccf.dat" -s'+pctype+' -r"mi.dat" > pceval.log'
            print('Running %s'%(cmd))
            os.system(cmd)
            ydata=np.loadtxt('ydata.dat')

    elif model == 'spp':

        ord=func_params[0]
        sp=func_params[1]
        pctype=func_params[2]

        np.savetxt('xdata.dat',xdata)

        npc=getNPC(dim,ord)
        #print "npc = ", npc
        # All coefs set to zero
        sppc=np.zeros((npc))

        # Randomly select some coefficients to be equal to 1
        nzind=np.random.choice(npc, sp,replace=False)

        for i in nzind:
            sppc[i]=1.0
        # Save the 'true' coefficients
        np.savetxt('sppc.dat',sppc)

        # Evaluate PC to get the training data
        cmd='pce_eval -x"PC" -o' + str(ord) + ' -f"sppc.dat" -s'+pctype+' > pceval.log'
        os.system(cmd)
        ydata=np.loadtxt('ydata.dat')

    elif model == 'currinExp':
    	# Currin, C., Mitchell, T., Morris, M., & Ylvisaker, D. (1988).
    	# A Bayesian approach to the design and analysis of computer experiments.
    	# Technical Report 6498. Oak Ridge National Laboratory.
    	# https://www.sfu.ca/~ssurjano/curretal88exp.html (March 2018)
    	xdata=0.5*(xdata+1.)
    	f1 = 1 - np.exp(-1/(2*xdata[:,1]));
    	f2 = 2300*xdata[:,0]**3 + 1900*xdata[:,0]**2 + 2092*xdata[:,0] + 60.0;
    	f3 = 100*xdata[:,0]**3 + 500*xdata[:,0]**2 + 4*xdata[:,0] + 20.0;
    	ydata = f1 * f2 / f3
    	return(ydata)

    elif model == 'currinExpLF':
        # see above comment on 'currinExp', and below
        # Xiong, S., Qian, P. Z., & Wu, C. J. (2013). Sequential design and analysis
        # of high-accuracy and low-accuracy computer codes. Technometrics, 55(1), 37-46
        xdataLF = xdata.copy()+0.1
        y1 = func(xdataLF,'currinExp',[]);
        for i in range(sam):
            xdataLF[i,1] = max(0.0,xdataLF[i,1]-0.2)
        y2 = func(xdataLF,'currinExp',[])
        xdataLF      = xdata.copy()-0.1
        xdataLF[:,1] = xdataLF[:,1]+0.2
        y3 = func(xdataLF,'currinExp',[])
        for i in range(sam):
            xdataLF[i,1] = max(0.0,xdataLF[i,1]-0.2)
        y4 = func(xdataLF,'currinExp',[])
        ydata = 0.25 * (y1 + y2 + y3 + y4)
        return ydata

    elif model == 'park91F1':
        # Park, J.S. (1991). Tuning complex computer codes to data and optimal designs.
        # Ph.D. Thesis, University of Illlinois, Champaign-Urbana.
        # https://www.sfu.ca/~ssurjano/park91a.html (March 2018)
        xdata=0.5*(xdata+1.)
        f1 = 0.5*xdata[:,0]*(np.sqrt(1.0 + (xdata[:,1]+xdata[:,2]**2)*xdata[:,3]/(xdata[:,0]**2)) - 1.0)
        f2 = (xdata[:,0] + 3*xdata[:,3]) * np.exp(1.0 + np.sin(xdata[:,2]))
        ydata = f1 + f2
        return ydata

    elif model == 'park91F1LF':
        # see above comment on 'park91F1', and below
        # Xiong, S., Qian, P. Z., & Wu, C. J. (2013). Sequential design and analysis
        # of high-accuracy and low-accuracy computer codes. Technometrics, 55(1), 37-46
        yHF=func(xdata,'park91F1',[])
        xdata=0.5*(xdata+1.)
        f1 = (1.0+0.1*np.sin(xdata[:,0])) * yHF
        f2 = -2*xdata[:,0] + xdata[:,1]**2 + xdata[:,2]**2
        ydata = f1 + f2 + 0.5;
        return ydata

    elif model == 'park91F2':
        # Park, J.S. (1991). Tuning complex computer codes to data and optimal designs.
        # Ph.D. Thesis, University of Illlinois, Champaign-Urbana.
        # https://www.sfu.ca/~ssurjano/park91a.html (March 2018)
        xdata=0.5*(xdata+1.)
        ydata = 2.0/3.0 * np.exp(xdata[:,0]+xdata[:,1]) - xdata[:,3] * np.sin(xdata[:,2]) + xdata[:,2]
        return ydata

    elif model == 'park91F2LF':
        # see above comment on 'park91F2', and below
        # Xiong, S., Qian, P. Z., & Wu, C. J. (2013). Sequential design and analysis
        # of high-accuracy and low-accuracy computer codes. Technometrics, 55(1), 37-46
        ydata = 1.2*func(xdata,'park91F2',[])-1.0
        return ydata

    else:
        print('Function type is not recognized. Exiting.')
        sys.exit()


    return ydata

##################################################################################

def integ_exact(model,func_params):
    """Analytically available function integrals.
    Note:
        * Currently only genz functions are implemented.
        * Note that conventional Genz arguments are in [0,1], here the expected input is on [-1,1]
    Arguments:
        * model       : Model name, options are 'genz_osc', 'genz_exp', 'genz_cont', 'genz_gaus', 'genz_cpeak', 'genz_ppeak'
        * func_params : Auxiliary parameters
                      : For genz functions, an array of size d+1, the first entry being the shift, and the rest of the entries are the weights.
                      : See UQTk Manual for Genz integral formulae.
    Returns:
        * integ_ex    : A real number that is the integral over [-1,1]^d
    """


    if (model=='genz_osc'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=cos(2.*pi*gcf[0]+0.5*sum(gcf[1:]))
        for i in range(1,dim+1):
            integ_ex*=(2.*sin(gcf[i]/2.)/gcf[i])
    elif (model=='genz_exp'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=1.
        for i in range(1,dim+1):
            at1=exp(-gcf[i]*gcf[0])
            at2=exp(gcf[i]*(1.-gcf[0]))
            integ_ex*=((at2-at1)/(gcf[i]))
    elif (model=='genz_cont'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=1.
        for i in range(1,dim+1):
            integ_ex*= ((2.-exp(gcf[i]*(-gcf[0]))-exp(gcf[i]*(gcf[0]-1.)))/gcf[i])
    elif (model=='genz_gaus'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=1.
        for i in range(1,dim+1):
            at1=erf(-gcf[i]*gcf[0])
            at2=erf(gcf[i]*(1.-gcf[0]))
            integ_ex*=((at2-at1)*sqrt(pi)/(2.*gcf[i]))
    elif (model=='genz_cpeak'):
        gcf=func_params
        dim=gcf.shape[0]-1
        numer=0.0
        count=1
        denom=1.
        for i in range(1,dim+1):
            comb=list(itertools.combinations(range(1,dim+1),i))
            for j in range(len(comb)):
                assert(i==len(comb[j]))
                #print i,j,pow(-1,i)
                numer+=(pow(-1,i)/(1.+sum(gcf[list(comb[j])])))
                count+=1
            denom*=(i*gcf[i])
        #print count, numer
        integ_ex=(1.+numer)/denom
    elif (model=='genz_ppeak'):
        gcf=func_params
        dim=gcf.shape[0]-1
        integ_ex=1.
        for i in range(1,dim+1):
            at1=np.arctan(-gcf[i]*gcf[0])
            at2=np.arctan(gcf[i]*(1.-gcf[0]))
            integ_ex*=(gcf[i]*(at2-at1))

    return integ_ex

################################################################################
################################################################################

def mainsens_exact(model,func_params):
    """Analytically available main sensitivities for some functions.
    Note:
        * Currently only sobol, ishigami and poly_exsens functions are implemented.
        * Note that conventional sobol arguments are in [0,1], here the expected input is on [-1,1]
    Arguments:
        * model       : Model name, options are 'sobol', 'ishigami', 'poly_exsens'
        * func_params : Auxiliary parameters
    Returns:
        * mainsens    : Main effect Sobol sensitivity index
    """
    if (model=='sobol'):
        dim=func_params.shape[0]
        mainsens=np.empty((dim,))
        var=1.0
        for i in range(dim):
            mainsens[i]=1./(3.*(1.+func_params[i])**2)
            var*=(mainsens[i]+1.)
        var-=1.0
        mainsens/=var

    elif (model=='ishigami'):
        a=func_params[0]
        b=func_params[1]
        var=a**2/8.+b*np.pi**4/5.+b**2*np.pi**8/18.+0.5
        mainsens=np.empty((3,))
        mainsens[0]=b*np.pi**4/5.+b**2*np.pi**8/50.+0.5
        mainsens[1]=a**2/8.
        mainsens[2]=0.0
        mainsens/=var

    elif (model=='poly_exsens'):
        dim=func_params[0]
        mainsens=(0.2/(1.2**dim-1))*np.ones((dim,))

    else:
        print('No exact sensitivity available for this function. Exiting.')
        sys.exit(1)


    return mainsens

##################################################################################
##################################################################################

def main(arg):
    modelname=arg[0]
    input_file=arg[1]
    output_file=arg[2]
    auxparam=[]
    if len(arg)>3:
        auxparam_file=arg[3]
        auxparam=np.loadtxt(auxparam_file,ndmin=1)

    input=np.loadtxt(input_file,ndmin=2)

    output=func(input,modelname,auxparam)
    np.savetxt(output_file,output)

if __name__ == "__main__":
    main(sys.argv[1:])

File: cpp/lib/tools/func.cpp
/* =====================================================================================

                      The UQ Toolkit (UQTk) version 3.1.5
                          Copyright (2024) NTESS
                        https://www.sandia.gov/UQToolkit/
                        https://github.com/sandialabs/UQTk

     Copyright 2024 National Technology & Engineering Solutions of Sandia, LLC (NTESS).
     Under the terms of Contract DE-NA0003525 with NTESS, the U.S. Government
     retains certain rights in this software.

     This file is part of The UQ Toolkit (UQTk)

     UQTk is open source software: you can redistribute it and/or modify
     it under the terms of BSD 3-Clause License

     UQTk is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     BSD 3 Clause License for more details.

     You should have received a copy of the BSD 3 Clause License
     along with UQTk. If not, see https://choosealicense.com/licenses/bsd-3-clause/.

     Questions? Contact the UQTk Developers at https://github.com/sandialabs/UQTk/discussions
     Sandia National Laboratories, Livermore, CA, USA
===================================================================================== */
/// \file func.cpp
/// \brief Implements several functions of form \f$y=f(\lambda;x)\f$

#include <math.h>
#include <cfloat>
#include <assert.h>

#include "func.h"

#include "gen_defs.h"
#include "PCSet.h"
#include "error_handlers.h"
#include "arrayio.h"
#include "arraytools.h"


// Proportionality function f(p,x) = p * x
Array2D<double> Func_Prop(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==1);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=pp(ip,0)*x(ix,0);

	return y;
}

// Quadratic proportionality f(p,x) = p1 * x + p2 * x^2
Array2D<double> Func_PropQuad(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==2);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=pp(ip,0)*x(ix,0)+pp(ip,1)*x(ix,0)*x(ix,0);

	return y;
}

// Exponential function f(p,x) = exp ( p1 + p2 * x )
Array2D<double> Func_Exp(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==2);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=exp(pp(ip,0)+pp(ip,1)*x(ix,0));

	return y;
}

// Quadratic exponential function f(p,x) = exp ( p1 + p2 * x + p3 * x^2 )
Array2D<double> Func_ExpQuad(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==3);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=exp(pp(ip,0)+pp(ip,1)*x(ix,0)+pp(ip,2)*x(ix,0)*x(ix,0));

	return y;
}

// Constant function f(p,x) = p
Array2D<double> Func_Const(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=pp(ip,0);

	return y;
}

// Linear function f(p,x) = p1 + p2 * x
Array2D<double> Func_Linear(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==2);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=pp(ip,0)+pp(ip,1)*x(ix,0);

	return y;
}

// Black-box function f(p,x), expects a script bb.x which reads p.dat and x.dat and writs y.dat
Array2D<double> Func_BB(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	Array2D<double> y;

	write_datafile(pp,"p.dat"); //np,pdim
    write_datafile(x,"x.dat"); //nx,xdim
    system("./bb.x");
    read_datafileVS(y,"y.dat"); //np,nx

    CHECKEQ(y.XSize(),np);
    CHECKEQ(y.YSize(),nx);

	return y;
}

// Heat transfer example function f(p,x) = T0 + (x * dw) / (Aw * p)
Array2D<double> Func_HT1(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==1);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	double To=273.;
   	double Aw=0.04;
    double dw=0.1;
	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=x(ix,0)*dw/(Aw*pp(ip,0))+To;

	return y;
}

// Heat transfer example function f(p,x) = p2 + (x * Q) / (Aw * p1)
Array2D<double> Func_HT2(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==2);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	double Aw=0.04;
    //double kw=1.0;
    double Q=20.0;
	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
		    y(ip,ix)=x(ix,0)*Q/(Aw*pp(ip,0))+pp(ip,1);

	return y;
}

// Fractional power example function f(p,x) = p1 + p2 * x + p3 * x^2 + p4 * (x + 1)^3.5
Array2D<double> Func_FracPower(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==4);
	assert(xdim==1);


	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=pp(ip,0)+pp(ip,1)*x(ix,0)+pp(ip,2)*pow(x(ix,0),2.)+pp(ip,3)*pow(x(ix,0)+1.,3.5);

	return y;
}

// Exponential example function for quick sketch f(p,x) = p2 * exp( x * p1 ) - 2
Array2D<double> Func_ExpSketch(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==2);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=pp(ip,1)*exp(x(ix,0)*pp(ip,0))-2.0;

	return y;
}

// Function that returns all inputs as a vector f(p,x_i) = p_i for i=1,2,..,nx
Array2D<double> Func_Inputs(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	assert(pdim==nx);
	assert(xdim==1);

	Array2D<double> y(np,nx);

	for(int ip=0;ip<np;ip++)
		for(int ix=0;ix<nx;ix++)
			y(ip,ix)=pp(ip,ix);

	return y;
}

// LU expansion f(p,x) = \sum_k p_k \Psi_k(x)
Array2D<double> Func_PCl(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	Array2D<int> mindexx;

	string pctype="LU";
    read_datafileVS(mindexx,"mindexx.dat");

    // Size checks
    CHECKEQ(mindexx.XSize(),pdim);
    CHECKEQ(mindexx.YSize(),xdim);


	PCSet surrmodel("NISPnoq",mindexx,pctype);

    Array2D<double> pcinput=x;

    Array2D<double> y(np,nx);

    for(int ip=0;ip<np;ip++){
    	Array1D<double> cf_this;
    	getRow(pp,ip,cf_this);
        Array1D<double> moutput;
        surrmodel.EvalPCAtCustPoints(moutput,pcinput,cf_this);

        for(int ix=0;ix<nx;ix++)
            y(ip,ix)=moutput(ix);
    }

	return y;
}

// LU expansion f(p,x) = \sum_k c_k \Psi(p,x)
Array2D<double> Func_PCx(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	Array2D<int> mindexpx;
	string pctype="LU";
    read_datafileVS(mindexpx,"mindexpx.dat");

    CHECKEQ(mindexpx.YSize(),pdim+xdim);

    Array1D<double> pccfpx(mindexpx.XSize());
    read_datafile_1d(pccfpx,"pccfpx.dat");

	PCSet surrmodel("NISPnoq",mindexpx,pctype);

    Array2D<double> pcinput(nx*np,pdim+xdim);
    for(int ip=0;ip<np;ip++){
	    for(int ix=0;ix<nx;ix++){
            for(int j=0;j<pdim;j++)
                pcinput(ip+ix*np,j)=pp(ip,j);
            for(int j=0;j<xdim;j++)
                pcinput(ip+ix*np,j+pdim)=x(ix,j);
        }
    }
    Array1D<double> model_samples_planar;
    surrmodel.EvalPCAtCustPoints(model_samples_planar,pcinput,pccfpx);

	Array2D<double> y(np,nx);

    for (int ip=0;ip<np;ip++)
        for (int ix=0;ix<nx;ix++)
            y(ip,ix)=model_samples_planar(ip+ix*np);

	return y;
}

// LU expansion f(p,x_i) = \sum_k c_k \Psi_k(p_i) when the multiindex is shared for all i
Array2D<double> Func_PC(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
	int* pred_mode=(int*) funcinfo;

	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	Array2D<int> mindexp;

	string pctype="LU";
    read_datafileVS(mindexp,"mindexp.dat");
    CHECKEQ(mindexp.YSize(),pdim);

    Array2D<double> pccf_all(mindexp.XSize(),nx);
    if ((*pred_mode)==0)
	    read_datafile(pccf_all,"pccf_all.dat");
    else
		read_datafile(pccf_all,"pccf_all_pred.dat");

	PCSet surrmodel("NISPnoq",mindexp,pctype);

    Array2D<double> pcinput=pp;

    Array2D<double> y(np,nx);

    for(int ix=0;ix<nx;ix++){
        Array1D<double> pccf_this;
        getCol(pccf_all,ix,pccf_this);
        Array1D<double> moutput;
        surrmodel.EvalPCAtCustPoints(moutput,pcinput,pccf_this);

        for(int ip=0;ip<np;ip++)
            y(ip,ix)=moutput(ip);
    }


	return y;
}

// LU expansion f(p,x_i) = \sum_k c_k \Psi_k(p_i) when multiindex can be different for each i
Array2D<double> Func_PCs(Array2D<double>& p, Array2D<double>& x, Array2D<double>& fixindnom, void* funcinfo)
{
    int* pred_mode=(int*) funcinfo;

	int np=p.XSize();
	int nx=x.XSize();
	int xdim=x.YSize();

	Array2D<double> pp=augment(p,fixindnom);

	int pdim=pp.YSize();

	string pctype="LU";

	Array1D<Array2D<int> > mindices(nx);
    Array1D<Array1D<double> > pccfs(nx);

   	for (int ix=0;ix<nx;ix++){
        char filename[25];
        int nn;

        if ((*pred_mode)==0)
            nn=sprintf(filename,"mindexp.%d.dat",ix);
        else
            nn=sprintf(filename,"mindexp.%d_pred.dat",ix);
        read_datafileVS(mindices(ix), filename);
        assert(mindices(ix).YSize()==pdim);

        if ((*pred_mode)==0)
            nn=sprintf(filename,"pccfp.%d.dat",ix);
        else
            nn=sprintf(filename,"pccfp.%d_pred.dat",ix);
        pccfs(ix).Resize(mindices(ix).XSize());

        read_datafile_1d(pccfs(ix), filename);

	}



    Array2D<double> pcinput=pp;

    Array2D<double> y(np,nx);

    for(int ix=0;ix<nx;ix++){
		PCSet surrmodel("NISPnoq",mindices(ix),pctype);

        Array1D<double> moutput;
        surrmodel.EvalPCAtCustPoints(moutput,pcinput,pccfs(ix));

        for(int ip=0;ip<np;ip++)
            y(ip,ix)=moutput(ip);
    }


	return y;
}

// Augments a parameter matrix with 'fixed' columns given indices and nominal values of those
Array2D<double> augment(Array2D<double>& p, Array2D<double>& fixindnom){

   	int fdim=fixindnom.XSize();
	if (fdim>0){
	    quicksort3(fixindnom,0, fdim-1,0);
    }


    Array2D<double> pp=p;


   	for (int i=0;i<fdim;i++){
        Array1D<double> nomCol(pp.XSize(),fixindnom(i,1));
        pp.insertCol(nomCol,(int)fixindnom(i,0));
	}


	return pp;
}

